teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,331 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+    LockToDefault bool
+    
+    // Deprecated is whether the feature is deprecated
+    // +optional
+    Deprecated bool
+}
+```
+
+## Test Plan
+### Retain existing CI end-to-end testing matrix 
+We plan to retain the existing testing matrix for fields at `alpha`, `beta` and `stable`. With the addition of per-feature flags for new features, it would look like:
+
+| |enable-api-fields|Per feature flags|Integration tests|
+| -- | -- | -- | -- |
+|Opt-in stable| stable|Turn OFF all per feature flags|Run all stable e2e tests.|
+|Opt-in beta| beta|Turn ON all beta per-feature flags|Run all beta e2e tests.|
+|Opt-in alpha| alpha|Turn ON all per-feature flags (alpha and beta)|Run all alpha e2e tests.|
+
+
+## Additional CI tests
+### Additional Test Combinations
+We cannot test 2\*\*N combinations of per feature flags since that would be too time consuming.
+Therefore, we try to mimic the following scenarios.
+
+- Say, a cluster operator has set the cluster to run Stable features. Occasionally, they may want to turn on a feature that is not yet stable. 
+- Similarly if the cluster is running Beta features by default (i.e. all feature flags at a beta stability level are ON and enable-api-fields is set to “stable”), cluster operators may want to turn on an individual feature at an alpha level. 
+- Conversely, if the cluster is running all the features (all feature flags are ON and enable api fields is set to “alpha”), they may want to turn off individual features.
+
+
+
+||enable-api-fields|Per feature flags|Integration tests|
+| :- | :- | :- | :- |
+|Opt-in stable|Set to stable|<p>All feature flags are OFF by default.</p><p>Turn ON one feature flag at a time. </p>|Run a small number of e2e tests against N combinations. It is not feasible to run the entire e2e test suite against N combinations.|
+|Opt-in beta|Set to beta|Turn ON all beta per-feature flags by default.<br><br>Turn ON one per-feature (flag at an alpha stability level) at a time.|Run a small number of e2e tests against M combinations (where M <= N)|
+|Opt-in alpha|Set to alpha|Turn ON all per-feature flags by default.<br><br>TURN OFF one feature flag at a time.|Run a small number of e2e tests against N combinations.|
+
+### <a name="_tcstxie74non"></a>How many tests can we run in a reasonable amount of time?
+Based on a [recent PR](https://github.com/tektoncd/pipeline/pull/7032), our integration tests take between 26 mins (stable) → 33 mins (alpha). We don’t want to go beyond that. Based on [Feature flags testing matrix](https://docs.google.com/document/d/1r_MX9-mzRtdbfNQq5VC4guHb-tphA0WxWhlQIsusEEA/edit?resourcekey=0-RALry7-GaKn9i19UEaRnYg) benchmarking, the approximate time is:
+
+T = N<sub>pipelines</sub>\*N<sub>tasks</sub>\*N<sub>features</sub>\*6 s
+
+
+
+Assume: 
+
+N<sub>features</sub> = 20 (currently we have 17 api features; lets assume that at any point in time, we will likely have ~20 individual features)
```

Comment:
This sounds a little disappointing, we need to do a better job of promoting or deprecating features. But it is reasonable given the maturity of the project.